numpy/core/src/umath/dispatching.c

```diff
@@ -0,0 +1,697 @@
+ *        WARNING: Returns a borrowed reference!
+ * @returns -1 on error 0 on success.  Note that the output can be NULL on
+ *          success if nothing is found.
+ */
+static int
+resolve_implementation_info(PyUFuncObject *ufunc,
+        PyArray_DTypeMeta *op_dtypes[], PyObject **out_info)
+{
+    int nin = ufunc->nin, nargs = ufunc->nargs;
+    /* Use new style type resolution has to happen... */
+    Py_ssize_t size = PySequence_Length(ufunc->_loops);
+    PyObject *best_dtypes = NULL;
+    PyObject *best_resolver_info = NULL;
+
+    for (Py_ssize_t res_idx = 0; res_idx < size; res_idx++) {
+        /* Test all resolvers  */
+        PyObject *resolver_info = PySequence_Fast_GET_ITEM(
+                ufunc->_loops, res_idx);
+        PyObject *curr_dtypes = PyTuple_GET_ITEM(resolver_info, 0);
+        /*
+         * Test if the current resolver matches, it could make sense to
+         * reorder these checks to avoid the IsSubclass check as much as
+         * possible.
+         */
+
+        npy_bool matches = NPY_TRUE;
+        /*
+         * NOTE: We check also the output DType.  In principle we do not
+         *       have to strictly match it (unless it is provided by the
+         *       `signature`).  This assumes that a (fallback) promoter will
+         *       unset the output DType if no exact match is found.
+         */
+        for (Py_ssize_t i = 0; i < nargs; i++) {
+            PyArray_DTypeMeta *given_dtype = op_dtypes[i];
+            PyArray_DTypeMeta *resolver_dtype = (
+                    (PyArray_DTypeMeta *)PyTuple_GET_ITEM(curr_dtypes, i));
+            assert((PyObject *)given_dtype != Py_None);
+            if (given_dtype == NULL && i >= nin) {
+                /* Unspecified out always matches (see below for inputs) */
+                continue;
+            }
+            if (given_dtype == resolver_dtype) {
+                continue;
+            }
+            if (!resolver_dtype->abstract) {
+                matches = NPY_FALSE;
+                break;
+            }
+            if (given_dtype == NULL) {
+                /*
+                 * If the (input) was not specified, this is a reduce-like
+                 * operation.  Some ufuncs may have non-trivial promotion
+                 * (e.g. add/multiply ensure high precision).
+                 * Continuing here matches promoters: those can deal with that.
+                 * If we allow this path for ArrayMethod, the person
+                 * registering will have it works sense for the ufunc, a
+                 * counter example is `(BoolLike, Bool, Bool)` for `add`.
+                 * It should resolve to an integer result (sum the bools)
+                 * in a reduction. But the ArrayMethod cannot work with that
+                 * (NumPy will prevent it to ensure correctness).
+                 */
+                continue;
+            }
+            int subclass = PyObject_IsSubclass(
```

Comment:
It seems strange that this line would not be covered. Is it true? Is it because we are only testing with numpy types right now?

Reply:
The whole subclass logic makes only sense if we have abstract DTypes.  Those do not really exist yet.  (Some do exist, but they only ever take part in value-based promotion, and currently always uses the legacy fallback always.)

It will be used if we have both:
* Promoters
* DTypes beyond "legacy" ones.